### PR TITLE
Changing 1-to-M behaviour of on_disk_cache.

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
TLDR: If filename_fn of on_disk_cache ( filename_fn'1 ) generates different name from filename_fn of end_caching, it is considered 1 to many situation. In this case additional listing file would be created at filename_fn'1 position and it will be used to check cache consistency between runs.

BC breaking, on_disk_cache no longer accept generators as `filename_fn`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #810

Differential Revision: [D40148560](https://our.internmc.facebook.com/intern/diff/D40148560)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40148560).

File: torchdata/datapipes/iter/util/cacheholder.py [hunk 1]
```diff
@@ -185,13 +194,19 @@ def __init__(
 
         if filepath_fn is not None:
             _check_unpickable_fn(filepath_fn)
-        filepath_fn = _generator_to_list(filepath_fn) if inspect.isgeneratorfunction(filepath_fn) else filepath_fn
+        assert not inspect.isgeneratorfunction(filepath_fn)  # BC breaking, now only str is accepted as return
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/cacheholder.py:197]: Then, we need to change the dataset implementation from TorchText side. e.g.: https://github.com/pytorch/text/blob/ff1fdfce8ac030a11638b2d94d54364144586253/torchtext/datasets/imdb.py#L107

File: torchdata/datapipes/iter/util/cacheholder.py [hunk 2]
```diff
@@ -206,66 +221,88 @@ def __add__(self, other_datapipe):
...
-                hash_type=hash_type,
-                extra_check_fn=extra_check_fn,
-            ),
+        filepath_fn, hash_dict, hash_type, extra_check_fn, cache_uuid = OnDiskCacheHolderIterDataPipe._temp_dict.pop(
+            self
         )
+
+        todo_dp: Any
+        cached_dp: Any
+        one_many_cached_dp: Any
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/cacheholder.py:276]: nit: annotated as IterDataPipe

**Review by ejguan — APPROVED** [review]
LGTM!!!! Thank you!!!

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40148560).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40148560).

**Pull request closed without merging**